### DISCUSSION TITLE
Broken Inlay hints on rust-analyzer

### DISCUSSION THREAD

**Discussion by milanjovanovic** [joaotavora/eglot#1239] [General]
I've noticed broken inlay hints on emacs29/eglot/rust-analyzer (macOS).
Couple of examples:

<img width="792" alt="broken_1" src="https://github.com/joaotavora/eglot/assets/3064500/a7edca5a-8484-4df3-83b6-61816090733a">
<img width="634" alt="broken_2" src="https://github.com/joaotavora/eglot/assets/3064500/42a2ee53-fbe1-4d47-adb3-873285e92cdc">

(Eldoc is showing correct type though)

**Comment by milanjovanovic** [comment]
This can be reproduced without .emacs file and without .emacs.d directory.
Just open rust project file and then M-x rust-ts-mode and eglot.

**Reply by joaotavora** [reply]
That's good, but to reproduce this I also need the project that you're using.  In the simple examples I try Rust in, I don't see this.  Is the code you're using open source?

**Reply by milanjovanovic** [reply]
I'm able to reproduce it with this code:

```
use std::error::Error;

struct Foo {}

struct Bar {}

type MyResult = Result<Foo, (Bar, String, Box<dyn Error + Send>)>;

fn get_vec_results() -> Vec<MyResult> {
    todo!()
}

fn main() {
    for result in get_vec_results() {
        match result {
            Ok(res) => todo!(),
            Err(err) => todo!(),
        }
    }
}
```



Screenshot from emacs 29.0.90 (no .emacs, no .emacs.d)
<img width="662" alt="broken_3" src="https://github.com/joaotavora/eglot/assets/3064500/a4f43db5-7ebb-43e6-8031-ebb403ee94b3">

**Reply by joaotavora** [reply]
I **cannot** reproduce with Emacs 30 built a few weeks ago.  Will try with Emacs 29, but I'm skeptical.  For reference here's the recipe I used:

```
$ emacs --version
GNU Emacs 30.0.50
Development version 8ce6e4c8293f on master branch; build date 2023-06-08.
Copyright (C) 2023 Free Software Foundation, Inc.
GNU Emacs comes with ABSOLUTELY NO WARRANTY.
You may redistribute copies of GNU Emacs
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING.
$ rust-analyzer --version
rust-analyzer 1 (2f1b7cedcf 2023-06-04)
$ emacs -Q src/main.rs -f rust-ts-mode -f eglot
```

**Reply by joaotavora** [reply]
I did spot a minor bug, which is you have to do some change in the buffer to see the inlay hints appear after Eglot connects.

**Reply by joaotavora** [reply]
Also can't reproduce with Emacs 29.  See https://joaotavora.github.io/eglot/#Troubleshooting-Eglot for a a systematic way to report Eglot bugs in a way that eliminates many of the unknown variables.  Maybe you're using some newer, or older `rust-analyzer` version?

**Reply by milanjovanovic** [reply]
I just tried couple of versions on both Intel and M2, I'm getting the same thing.

This is on Intel:
rust-analyzer 1.70.0 (90c54180 2023-05-31) (latest rustup version, also tried with older rust-analyzer-2023-04-03 version)
GNU Emacs 29.0.91 (build 1, x86_64-apple-darwin18.7.0 (this is from https://emacsformacosx.com, I've tried also with brew emacs29  (homebrew-emacs-plus)

/Applications/Emacs.app/Contents/MacOS/Emacs -Q src/main.rs -f rust-ts-mode -f eglot

I'm getting the same result as in image attached in previous comment.

I'm attaching eglot events buffer for both rust-analyzer versions.

[eglot.event.rust-analyzer.rustup.txt](https://github.com/joaotavora/eglot/files/11757761/eglot.event.rust-analyzer.rustup.txt)
[eglot.event.rust-analyzer-2023-04-03.txt](https://github.com/joaotavora/eglot/files/11757763/eglot.event.rust-analyzer-2023-04-03.txt)

**Reply by joaotavora** [reply]
May be a Mac OS thing?    Can you please report this to bug-gnu-emacs@gnu.org?  Keep the all-important Emacs -Q recipe and the rest of the information.  Additionally report how you (probably) use `cargo new` to create the example project (before you paste your code into `src/main.rs`)
